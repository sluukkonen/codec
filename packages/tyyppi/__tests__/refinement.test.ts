import { errors, refinement, string, unknown } from "../src/index.js"
import { expectParseFailure, expectParseSuccess } from "./helpers.js"

test("should parse successfully if the predicate returns true", () => {
  expectParseSuccess(
    refinement(
      unknown,
      () => true,
      () => {
        throw new Error("Boom!")
      },
    ),
    1,
  )
})

test("should fail to parse if the predicate returns false", () => {
  expectParseFailure(
    refinement(
      unknown,
      () => false,
      () => errors.custom({ message: "hmm?" }),
    ),
    1,
  )
})

test("should fail to parse if the underlying codec fails to parse the value", () => {
  expectParseFailure(
    refinement(
      string,
      () => true,
      () => {
        throw new Error("Boom!")
      },
    ),
    1,
  )
})

test("should allow configuring the metadata", () => {
  const meta = { tag: "hmm", simple: true }
  const codec = refinement(
    string,
    () => true,
    () => errors.custom({ message: "hmm?" }),
    meta,
  )
  expect(codec.meta).toEqual(meta)
})
