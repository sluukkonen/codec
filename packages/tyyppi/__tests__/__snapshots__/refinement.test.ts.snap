// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should fail to parse if the predicate returns false 1`] = `
[
  {
    "code": "custom",
    "message": "hmm?",
    "params": {},
    "path": [],
  },
]
`;

exports[`should fail to parse if the underlying codec fails to parse the value 1`] = `
[
  {
    "code": "invalid_type",
    "expected": "string",
    "message": "Expected a string, received a number",
    "path": [],
    "received": "number",
  },
]
`;
